File: ui/sentry/scrubbers.test.js
import { describe, it, expect } from 'vitest'
import { scrubString, scrubUrl, beforeSend, beforeBreadcrumb } from './scrubbers.js'

describe('scrubString', () => {
  it('collapses absolute user paths to ~/', () => {
    expect(scrubString('/Users/vladyslav/Movies/x.mkv')).toBe('~/Movies/*.mkv')
    expect(scrubString('/home/vlad/video.mp4')).toBe('~/*.mp4')
  })

  it('strips media filename stems, keeping extension', () => {
    expect(scrubString('Failed to transcode The.Sopranos.S01E03.mkv'))
      .toBe('Failed to transcode *.mkv')
    expect(scrubString('movie.MP4')).toBe('*.mp4')
  })

  it('is case-insensitive for media extensions', () => {
    expect(scrubString('/Users/a/x.MKV')).toBe('~/*.mkv')
  })

  it('redacts TVMaze/IMDb search terms', () => {
    expect(scrubString('Failed to fetch TVMaze: Sopranos'))
      .toBe('Failed to fetch TVMaze: <redacted>')
    expect(scrubString('Failed to fetch IMDb search: The Matrix'))
      .toBe('Failed to fetch IMDb search: <redacted>')
  })

  it('leaves non-sensitive strings unchanged', () => {
    expect(scrubString('ECONNREFUSED 127.0.0.1:3001'))
      .toBe('ECONNREFUSED 127.0.0.1:3001')
  })

  it('handles non-string input by returning it unchanged', () => {
    expect(scrubString(undefined)).toBe(undefined)
    expect(scrubString(null)).toBe(null)
    expect(scrubString(42)).toBe(42)
  })

  it('handles usernames with spaces', () => {
    expect(scrubString('/Users/Foo Bar/Movies/x.mkv')).toBe('~/Movies/*.mkv')
    expect(scrubString('/home/jane doe/video.mp4')).toBe('~/*.mp4')
  })
})

describe('scrubUrl', () => {
  it('replaces numeric id segments with :id', () => {
    expect(scrubUrl('/api/series/42/episodes/7'))
      .toBe('/api/series/:id/episodes/:id')
  })

  it('replaces UUID segments with :id', () => {
    expect(scrubUrl('/session/3f8e8a41-2b4c-4d5e-9f0a-1b2c3d4e5f6a/start'))
      .toBe('/session/:id/start')
  })

  it('strips query strings entirely', () => {
    expect(scrubUrl('/search?q=sopranos&page=2')).toBe('/search')
  })

  it('preserves non-id path segments', () => {
    expect(scrubUrl('/api/health')).toBe('/api/health')
  })

  it('handles absolute URLs', () => {
    expect(scrubUrl('http://localhost:3001/api/series/42?t=1'))
      .toBe('http://localhost:3001/api/series/:id')
  })

  it('handles non-string input by returning it unchanged', () => {
    expect(scrubUrl(undefined)).toBe(undefined)
    expect(scrubUrl(null)).toBe(null)
  })
})

describe('beforeSend', () => {
  it('scrubs message, exception value, stack filenames, request url', () => {
    const event = {
      message: 'Failed to transcode /Users/vladyslav/Movies/x.mkv',
      exception: {
        values: [
          {
            value: 'Cannot read /Users/vladyslav/a.mkv',
            stacktrace: {
              frames: [
                { filename: '/Users/vladyslav/code/caramba/web/src/App.jsx' },
              ],
            },
          },
        ],
      },
      request: { url: 'http://localhost:3001/api/series/42?t=1' },
    }
    const result = beforeSend(event)
    expect(result.message).toBe('Failed to transcode ~/Movies/*.mkv')
    expect(result.exception.values[0].value).toBe('Cannot read ~/*.mkv')
    expect(result.exception.values[0].stacktrace.frames[0].filename)
      .toBe('~/code/caramba/web/src/App.jsx')
    expect(result.request.url).toBe('http://localhost:3001/api/series/:id')
  })

  it('returns the same event object (mutates in place is fine)', () => {
    const event = { message: 'ok' }
    expect(beforeSend(event)).toBe(event)
  })

  it('tolerates missing optional fields', () => {
    expect(beforeSend({})).toEqual({})
  })

  it('clears event.request.query_string', () => {
    const event = {
      request: { url: '/api/x?t=1', query_string: 'q=sopranos&page=2' },
    }
    beforeSend(event)
    expect(event.request.url).toBe('/api/x')
    expect(event.request.query_string).toBe('')
  })

  it('scrubs frame abs_path alongside filename', () => {
    const event = {
      exception: { values: [{
        stacktrace: { frames: [
          { filename: 'app:///index.js', abs_path: '/Users/vladyslav/app.js' },
        ]},
      }]},
    }
    beforeSend(event)
    expect(event.exception.values[0].stacktrace.frames[0].abs_path)
      .toBe('~/app.js')
  })

  it('deletes event.user.username', () => {
    const event = { user: { username: 'vladyslav', id: '1' } }
    beforeSend(event)
    expect(event.user.username).toBeUndefined()
    expect(event.user.id).toBe('1')
  })
})

describe('beforeBreadcrumb', () => {
  it('scrubs message and data.url and data.to', () => {
    const crumb = {
      message: 'Navigation to /series/42',
      data: {
        url: 'http://localhost:3001/api/series/42?t=1',
        to: '/series/42/episode/7',
      },
    }
    const result = beforeBreadcrumb(crumb)
    expect(result.message).toBe('Navigation to /series/:id')
    expect(result.data.url).toBe('http://localhost:3001/api/series/:id')
    expect(result.data.to).toBe('/series/:id/episode/:id')
  })

  it('scrubs data.from on navigation breadcrumbs', () => {
    const crumb = { data: { from: '/series/42', to: '/series/43' } }
    beforeBreadcrumb(crumb)
    expect(crumb.data.from).toBe('/series/:id')
    expect(crumb.data.to).toBe('/series/:id')
  })

  it('tolerates missing data', () => {
    expect(beforeBreadcrumb({ message: 'x' })).toEqual({ message: 'x' })
  })
})

File: server/test/lib/sentry/scrubbers_test.rb
require "test_helper"
require "sentry/scrubbers"

class Sentry::ScrubbersTest < ActiveSupport::TestCase
  # ── scrub_string ─────────────────────────────────────────────────

  test "collapses absolute POSIX user paths to ~/" do
    assert_equal "~/Movies/*.mkv", Sentry::Scrubbers.scrub_string("/Users/vladyslav/Movies/x.mkv")
    assert_equal "~/*.mp4",         Sentry::Scrubbers.scrub_string("/home/vlad/video.mp4")
  end

  test "strips media filename stems, keeping lowercase extension" do
    assert_equal "Failed to transcode *.mkv", Sentry::Scrubbers.scrub_string("Failed to transcode The.Sopranos.S01E03.mkv")
    assert_equal "*.mp4", Sentry::Scrubbers.scrub_string("movie.MP4")
  end

  test "handles usernames with spaces" do
    assert_equal "~/Movies/*.mkv", Sentry::Scrubbers.scrub_string("/Users/Foo Bar/Movies/x.mkv")
  end

  test "redacts TVMaze and IMDb search terms" do
    assert_equal "Failed to fetch TVMaze: <redacted>", Sentry::Scrubbers.scrub_string("Failed to fetch TVMaze: Sopranos")
    assert_equal "Failed to fetch IMDb search: <redacted>", Sentry::Scrubbers.scrub_string("Failed to fetch IMDb search: The Matrix")
  end

  test "leaves non-sensitive strings unchanged" do
    assert_equal "ECONNREFUSED 127.0.0.1:3001", Sentry::Scrubbers.scrub_string("ECONNREFUSED 127.0.0.1:3001")
  end

  test "handles non-string input by returning it unchanged" do
    assert_nil Sentry::Scrubbers.scrub_string(nil)
    assert_equal 42, Sentry::Scrubbers.scrub_string(42)
  end

  # ── scrub_url ────────────────────────────────────────────────────

  test "replaces numeric id segments with :id" do
    assert_equal "/api/series/:id/episodes/:id", Sentry::Scrubbers.scrub_url("/api/series/42/episodes/7")
  end

  test "replaces UUID segments with :id" do
    assert_equal "/session/:id/start", Sentry::Scrubbers.scrub_url("/session/3f8e8a41-2b4c-4d5e-9f0a-1b2c3d4e5f6a/start")
  end

  test "strips query strings entirely" do
    assert_equal "/search", Sentry::Scrubbers.scrub_url("/search?q=sopranos&page=2")
  end

  test "preserves non-id path segments" do
    assert_equal "/api/health", Sentry::Scrubbers.scrub_url("/api/health")
  end

  test "handles absolute URLs" do
    assert_equal "http://localhost:3001/api/series/:id", Sentry::Scrubbers.scrub_url("http://localhost:3001/api/series/42?t=1")
  end

  test "returns non-strings unchanged" do
    assert_nil Sentry::Scrubbers.scrub_url(nil)
  end

  # ── before_send ──────────────────────────────────────────────────

  test "before_send scrubs message, exception value, frame filename and abs_path, request url and query_string" do
    event = {
      message: "Failed to transcode /Users/vladyslav/Movies/x.mkv",
      exception: {
        values: [
          {
            value: "Cannot read /Users/vladyslav/a.mkv",
            stacktrace: {
              frames: [
                { filename: "/Users/vladyslav/server/app/services/x.rb",
                  abs_path: "/Users/vladyslav/server/app/services/x.rb" },
              ],
            },
          },
        ],
      },
      request: { url: "http://localhost:3001/api/series/42?t=1", query_string: "t=1&q=sopranos" },
    }
    result = Sentry::Scrubbers.before_send(event)
    assert_equal "Failed to transcode ~/Movies/*.mkv", result[:message]
    assert_equal "Cannot read ~/*.mkv", result[:exception][:values][0][:value]
    assert_equal "~/server/app/services/x.rb", result[:exception][:values][0][:stacktrace][:frames][0][:filename]
    assert_equal "~/server/app/services/x.rb", result[:exception][:values][0][:stacktrace][:frames][0][:abs_path]
    assert_equal "http://localhost:3001/api/series/:id", result[:request][:url]
    assert_equal "", result[:request][:query_string]
  end

  test "before_send deletes user.username but preserves other user fields" do
    event = { user: { username: "vladyslav", id: "1" } }
    result = Sentry::Scrubbers.before_send(event)
    assert_nil result[:user][:username]
    assert_equal "1", result[:user][:id]
  end

  test "before_send tolerates missing fields" do
    assert_equal({}, Sentry::Scrubbers.before_send({}))
  end

  test "before_send returns the same event object (in-place mutation ok)" do
    event = { message: "ok" }
    assert_same event, Sentry::Scrubbers.before_send(event)
  end

  # ── before_breadcrumb ────────────────────────────────────────────

  test "before_breadcrumb scrubs message, url, from, and to" do
    crumb = {
      message: "Navigation to /series/42",
      data: {
        url: "http://localhost:3001/api/series/42?t=1",
        from: "/series/42",
        to: "/series/43/episode/7",
      },
    }
    result = Sentry::Scrubbers.before_breadcrumb(crumb)
    assert_equal "Navigation to /series/:id", result[:message]
    assert_equal "http://localhost:3001/api/series/:id", result[:data][:url]
    assert_equal "/series/:id", result[:data][:from]
    assert_equal "/series/:id/episode/:id", result[:data][:to]
  end

  test "before_breadcrumb tolerates missing data" do
    assert_equal({ message: "x" }, Sentry::Scrubbers.before_breadcrumb({ message: "x" }))
  end
end
